File: AdminService/src/main/java/com/ecommercial/shopping/adminservice/global/database/RoutingDataSource.java
package com.ecommercial.shopping.adminservice.global.database;

import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource;
import org.springframework.transaction.support.TransactionSynchronizationManager;
import java.util.List;
import java.util.Map;

public class RoutingDataSource extends AbstractRoutingDataSource {

    private static final String READ = "read";
    private static final String WRITE = "write";
    private final ReadOnlyDataSourceCycle<String> readOnlyDataSourceCycle = new ReadOnlyDataSourceCycle<>();


    @Override
    public void setTargetDataSources(Map<Object, Object> targetDataSources) {
        super.setTargetDataSources(targetDataSources);
        List<String> readOnlyDataSourceLookupKeys = targetDataSources.keySet()
                .stream()
                .map(String::valueOf)
                .filter(lookupKey -> lookupKey.contains(READ)).toList();
        readOnlyDataSourceCycle.setReadOnlyDataSourceLookupKeys(readOnlyDataSourceLookupKeys);
    }

    @Override
    protected Object determineCurrentLookupKey() {
        boolean isReadOnly = TransactionSynchronizationManager.isCurrentTransactionReadOnly();
        System.out.println(">>> [Routing] Transaction readOnly = " + isReadOnly);
        String lookupKey = isReadOnly
                ? readOnlyDataSourceCycle.getReadOnlyDataSourceLookupKey()
                : WRITE;
        System.out.println(">>> [Routing] Using DataSource = " + lookupKey);
        return lookupKey;
    }
}

File: UserService/src/main/kotlin/com/ecommercial/shopping/userservice/global/config/RoutingDataSource.kt
package com.ecommercial.shopping.userservice.global.config
import com.ecommercial.shopping.userservice.global.database.ReadOnlyDataSourceCycle
import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource
import org.springframework.transaction.support.TransactionSynchronizationManager

class RoutingDataSource : AbstractRoutingDataSource() {

    companion object {
        private const val READ = "read"
        private const val WRITE = "write"
    }

    private val readOnlyDataSourceCycle = ReadOnlyDataSourceCycle<String>()

    override fun setTargetDataSources(targetDataSources: MutableMap<Any, Any>) {
        super.setTargetDataSources(targetDataSources)
        val readOnlyDataSourceLookupKeys = targetDataSources.keys
            .map { it.toString() }
            .filter { it.contains(READ) }
        readOnlyDataSourceCycle.readOnlyDataSourceLookupKeys = readOnlyDataSourceLookupKeys
    }

    override fun determineCurrentLookupKey(): Any {
        val isReadOnly = TransactionSynchronizationManager.isCurrentTransactionReadOnly()
        println(">>> [Routing] Transaction readOnly = $isReadOnly")
        val lookupKey = if (isReadOnly) {
            readOnlyDataSourceCycle.getReadOnlyDataSourceLookupKey()
        } else {
            WRITE
        }
        println(">>> [Routing] Using DataSource = $lookupKey")
        return lookupKey
    }
}